cunumeric/vectorize.py

```diff
@@ -0,0 +1,643 @@
+            lines_to_skip = len(func.__doc__.split("\n"))
+
+        lines = inspect.getsourcelines(func)[0]  # type ignore
+
+        return_lines = []
+        for i in range(lines_to_skip + 1, len(lines)):
+            return_lines.append(lines[i].rstrip())
+        return return_lines
+
+    def _get_return_argumets(self) -> list[str]:
+        """
+        Returns the list of names for return arrays/values
+        """
+        self._func_body = self._get_func_body(self._pyfunc)
+        return_names = []
+        for ln in self._func_body:
+            if "return" in ln:
+                ln = ln.replace("return", "")
+                ln = ln.replace(" ", "")
+                return_names = ln.split(",")
+        return return_names
+
+    def _replace_name(
+        self, name: str, _LOOP_VAR: str, is_gpu: bool = False
+    ) -> str:
+        """
+        add indices to the names of input/output arrays in the function body
+        """
+        if (name in self._arg_names) or (name in self._return_names):
+            return "{}[int({})]".format(name, _LOOP_VAR)
+        else:
+            if is_gpu or ((not is_gpu) and not (name in self._scalar_names)):
+                return "{}".format(name)
+            else:
+                return "{}[0]".format(name)
+
+    def _build_gpu_function(self) -> Any:
+        funcid = "vectorized_{}".format(self._pyfunc.__name__)
+
+        # Preamble
+        lines = ["from numba import cuda"]
+        # we add math and numpy so user-defined functions can use them
+        lines.append("import math")
+        lines.append("import numpy")
+
+        # Signature
+        args = (
+            self._return_names
+            + self._arg_names
+            + self._scalar_names
+            + [_SIZE_VAR]
+            + [_DIM_VAR]
+            + [_PITCHES_VAR]
+            + [_STRIDES_VAR]
+        )
+
+        lines.append("def {}({}):".format(funcid, ",".join(args)))
+
+        # Initialize the index variable and return immediately
+        # when it exceeds the data size
+        lines.append("    local_i = cuda.grid(1)")
+        lines.append("    if local_i >= {}:".format(_SIZE_VAR))
+        lines.append("        return")
+        # we compute index for sparse data access when using Legion's
+        # pointer.
+        # a[x][y][z]=a[x*strides[0] + y*strides[1] + z*strides[2]]
+        lines.append("    {}:int = 0".format(_LOOP_VAR))
+        lines.append("    for p in range({}-1):".format(_DIM_VAR))
+        # FIXME make sure we compute index correct for all data types
+        lines.append("        x=int(local_i/{}[p])".format(_PITCHES_VAR))
+        lines.append(
+            "        local_i = int(local_i%{}[p])".format(_PITCHES_VAR)
+        )
+        lines.append(
+            "        {}+=int(x*{}[p])".format(_LOOP_VAR, _STRIDES_VAR)
+        )
+        lines.append(
+            "    {}+=int(local_i*{}[{}-1])".format(
+                _LOOP_VAR, _STRIDES_VAR, _DIM_VAR
+            )
+        )
```

Comment:
could all of these lines be assembled into one easily-readable triple-quoted format string with named format parameters? (this is a good instance to use `.format` rather than f-string) Then split here on newlines?